### PR TITLE
slice_*() careful about {{ }}

### LINKED ISSUE #6089: slice_max(order_by = n)
``` r
library(dplyr, warn.conflicts = FALSE)

df <- data.frame(x = 1:10, n = 10:1)
slice_max(df, order_by = n, n = 1)
#> Error: `order_by` must have size 10, not size 1.
```

<sup>Created on 2021-11-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

identified in https://github.com/tidyverse/dplyr/issues/6080 while testing `texter`

### PR DESCRIPTION
**Pull request by romainfrancois**
closes #6089

### CODE REVIEW

File: R/slice.R [hunk 1]
```diff
@@ -175,9 +175,9 @@ slice_min.data.frame <- function(.data, order_by, ..., n, prop, with_ties = TRUE
   }
 
   slice_impl(.data, {
-    n <- dplyr::n()
-    x <- vec_assert({{ order_by }}, size = n, arg = "order_by")
-    idx(x, n)
+    ...n <- dplyr::n()
```
Comment (romainfrancois) [inline R/slice.R:178]: or instead maybe this is better: 

```r
x <- vec_assert({{ order_by }}, size = n(), arg = "order_by")
idx(x, n())
```

the idea of having `n <- dplyr::n()` is that `n()` would only be called once.

File: R/slice.R [hunk 2]
```diff
@@ -203,9 +203,9 @@ slice_max.data.frame <- function(.data, order_by, ..., n, prop, with_ties = TRUE
...
+    x <- vec_assert({{ order_by }}, size = ...n, arg = "order_by")
+    idx(x, ...n)
```
Comment (DavisVaughan) [inline R/slice.R:208]:
```suggestion
    order_by <- {{ order_by }}
    n <- dplyr::n()
    order_by <- vec_assert(order_by, size = n, arg = "order_by")
    idx(order_by, n)
```

Would it work if the first thing you did was evaluate the quosure? Before any local variables like `n` are defined?

Could do the same approach with `weight_by` in `slice_sample.data.frame()` I think

Comment (DavisVaughan) [reply R/slice.R:208]: In my mind this is kind of like calling `force(order_by)`

Comment (romainfrancois) [reply R/slice.R]: Thanks. Thats' better.

**Comment by romainfrancois** [comment]
Actually that does not help in grouped situations: 

``` r
library(dplyr, warn.conflicts = FALSE)

tryCatch(error= print,
data.frame(x = 1:10, n = 1:10, g = rep(1:2, each = 5)) %>% 
  group_by(g) %>% 
  slice_max(order_by= n, n = 1)
)
#> <error/rlang_error>
#> Error in `slice_max()`: `order_by` must have size 5, not size 1.
#> ℹ The error occurred in group 2: g = 2.
#> Backtrace:
#>   1. base::tryCatch(...)
#>  17. dplyr `<fn>`(`<vctrs___>`)
```

<sup>Created on 2021-11-17 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

The size 1 thing this complains about is the `n` that was assigned from the previous group.

**Pull request merged**